src/datasets/packaged_modules/imagefolder/imagefolder.py

```diff
@@ -1,373 +1,50 @@
-                            )
-
-        data_files = self.config.data_files
-        splits = []
-        for split_name, files in data_files.items():
-            if isinstance(files, str):
-                files = [files]
-            files, archives = self._split_files_and_archives(files)
-            downloaded_files = dl_manager.download(files)
-            downloaded_dirs = dl_manager.download_and_extract(archives)
-            if do_analyze:  # drop_metadata is None or False, drop_labels is None or False
-                logger.info(f"Searching for labels and/or metadata files in {split_name} data files...")
-                analyze(files, downloaded_files, split_name)
-                analyze(archives, downloaded_dirs, split_name)
-
-                if metadata_files:
-                    # add metadata if `metadata_files` are found and `drop_metadata` is None (default) or False
-                    add_metadata = not (self.config.drop_metadata is True)
-                    # if `metadata_files` are found, add labels only if
-                    # `drop_labels` is set up to False explicitly (not-default behavior)
-                    add_labels = self.config.drop_labels is False
-                else:
-                    # if `metadata_files` are not found, don't add metadata
-                    add_metadata = False
-                    # if `metadata_files` are not found but `drop_labels` is None (default) or False, add them
-                    add_labels = not (self.config.drop_labels is True)
-
-                if add_labels:
-                    logger.info("Adding the labels inferred from data directories to the dataset's features...")
-                if add_metadata:
-                    logger.info("Adding metadata to the dataset...")
-            else:
-                add_labels, add_metadata, metadata_files = False, False, {}
-
-            splits.append(
-                datasets.SplitGenerator(
-                    name=split_name,
-                    gen_kwargs={
-                        "files": [(file, downloaded_file) for file, downloaded_file in zip(files, downloaded_files)]
-                        + [(None, dl_manager.iter_files(downloaded_dir)) for downloaded_dir in downloaded_dirs],
-                        "metadata_files": metadata_files,
-                        "split_name": split_name,
-                        "add_labels": add_labels,
-                        "add_metadata": add_metadata,
-                    },
-                )
+        # _prepare_split_generators() sets self.info.features,
+        # infers labels, finds metadata files if needed and returns splits
+        splits = self._prepare_split_generators(dl_manager)
+
+        add_labels = splits[0].gen_kwargs["add_labels"]
+        if self.config.features is None and add_labels:
+            task_template = ImageClassification(
+                image_column=self.config.base_feature_name, label_column=self.config.label_column
```

Comment:
Feel free to remove this stuff (the task template API is deprecated, although not officially). I guess this is why you defined the `_prepare_` methods in the first place.